program/src/processor.rs

```diff
@@ -5313,6 +5318,286 @@ impl Processor {
+
+        let [
+            mango_group_ai,         // read
+            mango_account_ai,       // write
+            owner_ai,               // read & signer
+            mango_cache_ai,         // read
+            signer_ai,              // read
+        ] = fixed_ais;
+
+        let mango_group = MangoGroup::load_checked(mango_group_ai, program_id)?;
+        check!(signer_ai.key == &mango_group.signer_key, MangoErrorCode::InvalidSignerKey)?;
+
+        let mut mango_account =
+            MangoAccount::load_mut_checked(mango_account_ai, program_id, mango_group_ai.key)?;
+        check!(owner_ai.is_signer, MangoErrorCode::SignerNecessary)?;
+        check!(
+            &mango_account.owner == owner_ai.key || &mango_account.delegate == owner_ai.key,
+            MangoErrorCode::InvalidOwner
+        )?;
+        check!(!mango_account.is_bankrupt, MangoErrorCode::Bankrupt)?;
+
+        // Take out the root bank, node bank and vault accounts and validate
+        let mut bank_sets = vec![];
+        let mut extra_user_assets = vec![];
+        let mut pre_amounts = vec![];
+        for bank_set_ais in bank_ais.chunks_exact(3) {
+            let bank_set_ais = array_ref![bank_set_ais, 0, 3];
+            let [root_bank_ai, node_bank_ai, vault_ai] = bank_set_ais;
+            let token_index = mango_group
+                .find_root_bank_index(root_bank_ai.key)
+                .ok_or(throw_err!(MangoErrorCode::InvalidRootBank))?;
+
+            let root_bank = RootBank::load_checked(root_bank_ai, program_id)?;
+
+            check!(
+                root_bank.node_banks.contains(node_bank_ai.key),
+                MangoErrorCode::InvalidNodeBank
+            )?;
+            let node_bank = NodeBank::load_mut_checked(node_bank_ai, program_id)?;
+            check_eq!(&node_bank.vault, vault_ai.key, MangoErrorCode::InvalidVault)?;
+            // pre_amounts.push(Account::unpack(&vault_ai.try_borrow_data()?)?.amount);
+            let token_account = TokenAccount::load_checked(vault_ai)?;
+            token_account.check_mango_reqs(&mango_group)?;
+
+            pre_amounts.push(token_account.0.amount);
+
+            bank_sets.push(BankSet { token_index, node_bank, vault_ai });
+            extra_user_assets.push((AssetType::Token, token_index));
+        }
+
+        // Validate MangoCache
+        let now_ts = Clock::get()?.unix_timestamp as u64;
+        let active_assets = UserActiveAssets::new(&mango_group, &mango_account, extra_user_assets);
+        let mango_cache = MangoCache::load_checked(mango_cache_ai, program_id, &mango_group)?;
+        mango_cache.check_valid(&mango_group, &active_assets, now_ts)?;
+
+        /**
+        There are two types of `Account`: either owned by Mango program id or not.
+        We are not worried about `Account` owned by Mango program because we disallow CPI into Mango
+            Therefore, Mango owned accounts cannot be modified
```

Comment:
Would it be possible to write a forwarder program that calls back into the mango program and that preserves the signatures?

Reply:
Solana says this type of reentrancy is not allowed. Something to be verified from Solana engineers and/or Neodyme

Reply:
https://docs.solana.com/developing/programming-model/calling-between-programs#reentrancy

Reply:
Great to see that this is guarded against in the runtime!